clients/drcachesim/tracer/tracer.cpp

```diff
@@ -918,8 +949,30 @@ instrument_memref(void *drcontext, user_data_t *ud, instrlist_t *ilist, instr_t
     }
     if (op_L0_filter.get_value())
         insert_load_buf_ptr(drcontext, ilist, where, reg_ptr);
-    adjust = instru->instrument_memref(drcontext, ilist, where, reg_ptr, adjust, app, ref,
-                                       ref_index, write, pred);
+
+#ifdef AARCH64
+    // TODO i#4400: Use a new operand type to back-align zeroed address, instead of
+    // doing it in drcachesim.
+    if (is_dc_zva_instr(app)) {
+        reg_id_t reg_ba_addr;
+        if (drreg_reserve_register(drcontext, ilist, where, NULL, &reg_ba_addr) !=
+            DRREG_SUCCESS) {
+            FATAL("Fatal error: failed to reserve register for back-aligned DC ZVA "
+                  "memory reference\n");
+        }
+        opnd_t back_aligned_memref = back_align_dc_zva_memref_opnd(
+            drcontext, ilist, where, ref, proc_get_cache_line_size(), reg_ba_addr);
+        adjust = instru->instrument_memref(drcontext, ilist, where, reg_ptr, adjust, app,
+                                           back_aligned_memref, ref_index, write, pred);
+
+        if (drreg_unreserve_register(drcontext, ilist, where, reg_ba_addr) !=
```

Comment:
For offline, this is all unnecessary extra register spilling and an and-ing: this is most optimally done in post-processing.  (If there were no runtime cost, then yes better to share the same code for online and offline; but given how common this instruction seems to be, we do not want to pay for all this extra runtime cost when it's so easy to avoid it).

In fact, given that this is supposed to be a temporary fix, and that our usage cares less about online, you could argue for putting the workaround *only* in post-processing and having online live with the bug until the proper fix is in place (no need for a separate .asm test then either).

Reply:
Oh okay, makes sense. As that implementation is completely different than this PR, I've opened a new one #4444 for that.